File: analyses/hepatic_fat_gwas_jan2026/make_manhattan_plot.R
library(data.table)
library(qqman)
library(dplyr)

# Paths
results_file <- "/home/agaro/verma_shared/projects/Liver_IDPs/analyses/hepatic_fat_gwas_manual_jan2026/PMBB_ALL_ALL/Sumstats/PMBB_ALL_ALL.hepatic_fat.gwas.saige.gz"
annot_file <- "/home/agaro/verma_shared/projects/Liver_IDPs/analyses/hepatic_fat_gwas_jan2026/Annotations/GWAS_biofilter_genes_rsids.csv"
output_plot <- "/home/agaro/verma_shared/projects/Liver_IDPs/analyses/hepatic_fat_gwas_manual_jan2026/Plots/hepatic_fat_manhattan_qqman.png"

# Load data
cat("Loading GWAS results...\n")
gwas <- fread(results_file)

# Load annotations
cat("Loading gene annotations...\n")
annot <- fread(annot_file)

# Merge
gwas <- merge(gwas, annot[, .(Var_ID, Gene, RSID)], 
              by.x = "variant_id", by.y = "Var_ID", 
              all.x = TRUE)

# Clean chromosome names
gwas$CHR <- as.integer(gsub("chr", "", gwas$chromosome))
gwas <- gwas[!is.na(CHR) & CHR %in% 1:22]

# Rename columns for qqman
gwas$BP <- gwas$base_pair_location
gwas$P <- gwas$p_value
gwas$SNP <- gwas$variant_id

# Remove missing p-values
gwas <- gwas[!is.na(P) & P > 0]

cat(sprintf("Total variants: %s\n", format(nrow(gwas), big.mark=",")))

# Separate significant and non-significant
sig_threshold <- 5e-8
sig <- gwas[P < sig_threshold]
nonsig <- gwas[P >= sig_threshold]

cat(sprintf("Significant variants: %s\n", nrow(sig)))
cat(sprintf("Non-significant variants: %s\n", nrow(nonsig)))

# Clump only significant variants (to avoid PNPLA3 repetition)
sig <- sig[order(CHR, BP, P)]
sig$region <- paste0(sig$CHR, "_", floor(sig$BP / 1e6))
sig_clumped <- sig %>%
  group_by(region) %>%
  slice_min(P, n = 1, with_ties = FALSE) %>%
  ungroup() %>%
  as.data.table()

cat(sprintf("Clumped significant to: %s independent signals\n", nrow(sig_clumped)))

# Thin non-significant variants for plotting speed (keep random 50k)
set.seed(42)
if (nrow(nonsig) > 50000) {
  nonsig_thinned <- nonsig[sample(.N, 50000)]
} else {
  nonsig_thinned <- nonsig
}

# Combine clumped significant + thinned non-significant
gwas_plot <- rbind(sig_clumped, nonsig_thinned, fill = TRUE)
gwas_plot <- gwas_plot[order(CHR, BP)]

# Add gene names for labeling
gwas_plot$LABEL <- ifelse(!is.na(gwas_plot$Gene) & gwas_plot$P < sig_threshold, 
                          gwas_plot$Gene, 
                          NA)

cat(sprintf("Total variants for plotting: %s\n", nrow(gwas_plot)))

# Create Manhattan plot
png(output_plot, width = 1800, height = 600, res = 100)

manhattan(gwas_plot,
         chr = "CHR",
         bp = "BP", 
         p = "P",
         snp = "LABEL",
         col = c("#4169E1", "#FFA500"),
         suggestiveline = -log10(1e-5),
         genomewideline = -log10(5e-8),
         annotatePval = 5e-8,
         annotateTop = TRUE,
         main = "GWAS Manhattan Plot: Hepatic Fat (N = 18,079)",
         ylim = c(0, max(-log10(gwas_plot$P)) * 1.05))

dev.off()

cat(sprintf("\nManhattan plot saved to: %s\n", output_plot))

File: analyses/hepatic_fat_gwas_jan2026/make_manhattan_plot.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os
from manhattan_plot import ManhattanPlot

# Configuration
cohort = "PMBB_ALL_ALL"
pheno = "hepatic_fat"

# Paths
results_dir = "/home/agaro/verma_shared/projects/Liver_IDPs/analyses/hepatic_fat_gwas_jan2026/PMBB_ALL_ALL/Sumstats"
annot_dir = "/home/agaro/verma_shared/projects/Liver_IDPs/analyses/hepatic_fat_gwas_jan2026/Annotations"
output_dir = "/home/agaro/verma_shared/projects/Liver_IDPs/analyses/hepatic_fat_gwas_jan2026/Plots"

sumstats_file = os.path.join(results_dir, f"{cohort}.{pheno}.gwas.saige.gz")
annot_file = os.path.join(annot_dir, "GWAS_biofilter_genes_rsids.csv")
pheno_table_file = "/home/agaro/verma_shared/projects/Liver_IDPs/analyses/hepatic_fat_gwas_jan2026/pheno_summaries.csv"  
colnames_file = "/home/agaro/verma_shared/projects/Liver_IDPs/analyses/hepatic_fat_gwas_jan2026/colnames.txt" 

# Load column mapping
print("Loading column mappings...")
columns_map = {}
with open(colnames_file, 'r') as file:
    for line in file:
        key, value = line.split('=')
        key = key.strip()
        value = value.strip().strip("'")
        columns_map[key] = value

# Reverse the columns map
columns_map_inv = {v: k for k, v in columns_map.items()}
print(f"Column mappings loaded: {columns_map}")

# Load phenotype data 
print("\nLoading phenotype information...")
pheno_df = pd.read_csv(pheno_table_file)
pheno_df = pheno_df[pheno_df['PHENO'] == pheno]
pheno_df = pheno_df[pheno_df['COHORT'] == cohort]


# Determine trait type 
trait_type = 'bin' if pheno_df['Cases'].count() != 0 else 'quant'
print(f"Trait type: {trait_type}")

# Set up output path
output_manhattan = os.path.join(output_dir, f'{cohort}.{pheno}.manhattan_vertical.png')
output_qq = os.path.join(output_dir, f'{cohort}.{pheno}.qq.png')

# Create output directory if it doesn't exist
os.makedirs(output_dir, exist_ok=True)

# Create plot title
plot_title = f'GWAS {cohort}: {pheno.replace("_", " ")}'

if trait_type == 'bin':
    plot_title += f'\nCases = {pheno_df.iloc[0].Cases:,.0f}, Controls = {pheno_df.iloc[0].Controls:,.0f}'
else:
    plot_title += f'\nN = {pheno_df.iloc[0].N:,.0f}'

print(f"\nPlot title: {plot_title}")

# Load and process data
print("\nInitializing ManhattanPlot...")
mp = ManhattanPlot(sumstats_file, test_rows=None, title=plot_title)
mp.load_data()

print("Processing chromosome and variant IDs...")
mp.df['chromosome_noCHR'] = mp.df['chromosome'].astype(str).str.replace('chr', '').astype(int)
mp.df['variant_id'] = mp.df['variant_id'].str.replace('_', ':')

print("Cleaning data with column mappings...")
mp.clean_data(col_map={
    'chromosome_noCHR': '#CHROM', 
    columns_map['POS']: 'POS', 
    columns_map['MarkerID']: 'ID', 
    columns_map['p.value']: 'P'
})

print(f"Data shape: {mp.df.shape}")
print(mp.df.head())

# Annotations
if annot_file is not None and os.path.exists(annot_file):
    print("\nLoading and adding annotations...")
    annot_df = pd.read_csv(annot_file)
    annot_df['ID'] = annot_df['Gene']
    annot_df.iloc[:, 0] = annot_df.iloc[:, 0].str.replace('_', ':')
    mp.add_annotations(annot_df, extra_cols=['RSID'])
    print(f"Annotations added: {len(annot_df)} entries")
else:
    print("\nNo annotation file found, skipping annotations...")


# Thin data 
print("\nThinning data for plotting...")
mp.get_thinned_data()
print(f"Thinned data points: {len(mp.thinned)}")
print(mp.thinned.head())

# Set annotation threshold
annot_thresh = 5E-8 if np.any(mp.thinned['P'].min() < 5E-8) else np.nanquantile(mp.thinned['P'], 10 / len(mp.thinned))
print(f"\nAnnotation threshold: {annot_thresh:.2e}")

# Update plotting parameters
print("Setting plotting parameters...")
mp.update_plotting_parameters(
    vertical=True, 
    sig=annot_thresh if not np.any(mp.thinned['P'] < 5E-8) else 5E-8, 
    sug=annot_thresh, 
    annot_thresh=annot_thresh, 
    merge_genes=True
)

# Create Manhattan plot
print("\nGenerating Manhattan plot...")
mp.full_plot(
    save=output_manhattan, 
    rep_boost=False, 
    extra_cols={'beta': 'BETA', 'RSID': 'RSID'}, 
    number_cols=['BETA'], 
    keep_chr_pos=False, 
    with_table_bg=False, 
    with_table_grid=False
)

plt.clf()
print(f"✓ Manhattan plot saved to: {output_manhattan}")

# Create QQ Plot 
print("\nGenerating QQ plot...")
mp.qq_plot(save=output_qq)
print(f"✓ QQ plot saved to: {output_qq}")

# Create plots manifest
print("\nCreating plots manifest...")

print("\n" + "="*50)
print("DONE! All plots generated successfully.")
print("="*50)